pkg/binding/copy_message.go

```diff
@@ -0,0 +1,26 @@
+package binding
+
+// CopyMessage reads m once and creates an in-memory copy using EventMessage or
+// StructMessage depending on the encoding of m.  The returned copy is not
+// dependent on any transport and can be read many times.
+//
+// NOTE: Calling Finish() on the returned message calls m.Finish()
+func CopyMessage(m Message) (Message, error) {
+	// Try structured first, it's cheaper.
+	sm := StructMessage{}
+	err := m.Structured(&sm)
+	switch err {
+	case nil:
+		return &sm, nil
+	case ErrNotStructured:
+		break
+	default:
+		return nil, err
+	}
+	em := EventMessage{}
+	err = m.Event(&em)
+	if err != nil {
+		return nil, err
+	}
+	return WithFinish(em, func(err error) { _ = m.Finish(err) }), nil
```

Comment:
Binding the original message lifecycle to the copied message is probably what you don't want to do if you want to decouple the lifecycles of the two messages (eg like you explained in this comment point 2 https://github.com/cloudevents/sdk-go/pull/282#issuecomment-574906328)

Reply:
Right, my gut tells me that in cases where you need to buffer for one reason (mult-write) you will probably need to buffer for others (retry after incoming message is gone etc.) so there's' limited value in optimizing the special case of buffer-for-multi-write-only as it probably won't come up much in reality. I could be wrong but lets fix that when there's a real use case and the creation of a single extra EventMessage really is the hot spot on the critical path.